// @angular-material-extensions/google-maps-autocomplete/bundles/angular-material-extensions-google-maps-autocomplete.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@agm/core'), require('@angular/animations'), require('@angular/common'), require('@angular/flex-layout'), require('@angular/material')) :
    typeof define === 'function' && define.amd ? define('@angular-material-extensions/google-maps-autocomplete', ['exports', '@angular/core', '@angular/forms', '@agm/core', '@angular/animations', '@angular/common', '@angular/flex-layout', '@angular/material'], factory) :
    (global = global || self, factory((global['angular-material-extensions'] = global['angular-material-extensions'] || {}, global['angular-material-extensions']['google-maps-autocomplete'] = {}), global.ng.core, global.ng.forms, global.core$1, global.ng.animations, global.ng.common, global.ng['flex-layout'], global.ng.material));
}(this, (function (exports, core, forms, core$1, animations, common, flexLayout, material) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/interfaces/germand.address.interface.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @author Anthony Nahas
     * \@since 23.12.19
     * @version 1.0
     * @record
     */
    function GermanAddress() { }
    if (false) {
        /** @type {?|undefined} */
        GermanAddress.prototype.id;
        /** @type {?} */
        GermanAddress.prototype.gmID;
        /** @type {?} */
        GermanAddress.prototype.placeID;
        /** @type {?|undefined} */
        GermanAddress.prototype.name;
        /** @type {?|undefined} */
        GermanAddress.prototype.icon;
        /** @type {?|undefined} */
        GermanAddress.prototype.displayAddress;
        /** @type {?|undefined} */
        GermanAddress.prototype.postalCode;
        /** @type {?|undefined} */
        GermanAddress.prototype.streetNumber;
        /** @type {?|undefined} */
        GermanAddress.prototype.streetName;
        /** @type {?|undefined} */
        GermanAddress.prototype.sublocality;
        /** @type {?|undefined} */
        GermanAddress.prototype.locality;
        /** @type {?|undefined} */
        GermanAddress.prototype.state;
        /** @type {?|undefined} */
        GermanAddress.prototype.country;
        /** @type {?|undefined} */
        GermanAddress.prototype.vicinity;
        /** @type {?|undefined} */
        GermanAddress.prototype.url;
        /** @type {?|undefined} */
        GermanAddress.prototype.geoLocation;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/interfaces/location.interface.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function Location() { }
    if (false) {
        /** @type {?} */
        Location.prototype.latitude;
        /** @type {?} */
        Location.prototype.longitude;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/interfaces/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/address-validator/mat-address-validator.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    // https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts
    var MatValidateAddressDirective = /** @class */ (function () {
        function MatValidateAddressDirective() {
        }
        /**
         * @return {?}
         */
        MatValidateAddressDirective.prototype.validate = /**
         * @return {?}
         */
        function () {
            var _this = this;
            return (/**
             * @param {?} control
             * @return {?}
             */
            function (control) {
                return _this.address ? null : {
                    validateAddress: {
                        valid: false
                    }
                };
            });
        };
        /**
         * @param {?} eventEmitter
         * @return {?}
         */
        MatValidateAddressDirective.prototype.subscribe = /**
         * @param {?} eventEmitter
         * @return {?}
         */
        function (eventEmitter) {
            var _this = this;
            this.subscription = eventEmitter.subscribe((/**
             * @param {?} address
             * @return {?}
             */
            function (address) {
                _this.address = address;
            }));
        };
        /**
         * @return {?}
         */
        MatValidateAddressDirective.prototype.unsubscribe = /**
         * @return {?}
         */
        function () {
            this.subscription.unsubscribe();
        };
        Object.defineProperty(MatValidateAddressDirective.prototype, "address", {
            get: /**
             * @return {?}
             */
            function () {
                return this._address;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                this._address = value;
            },
            enumerable: true,
            configurable: true
        });
        MatValidateAddressDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[mat-address-validate][formControlName],[MatValidateAddress][formControl],[MatValidateAddress][ngModel]',
                        providers: [
                            { provide: forms.NG_VALIDATORS, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return MatValidateAddressDirective; })), multi: true }
                        ]
                    },] }
        ];
        /** @nocollapse */
        MatValidateAddressDirective.ctorParameters = function () { return []; };
        return MatValidateAddressDirective;
    }());
    if (false) {
        /** @type {?} */
        MatValidateAddressDirective.prototype.subscription;
        /**
         * @type {?}
         * @private
         */
        MatValidateAddressDirective.prototype._address;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component/mat-google-maps-autocomplete.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var Appearance = {
        STANDARD: "standard",
        FILL: "fill",
        OUTLINE: "outline",
        LEGACY: "legacy",
    };
    var MatGoogleMapsAutocompleteComponent = /** @class */ (function () {
        function MatGoogleMapsAutocompleteComponent(mapsAPILoader, ngZone) {
            this.mapsAPILoader = mapsAPILoader;
            this.ngZone = ngZone;
            this.addressLabelText = 'Address';
            this.placeholderText = 'Please enter the address';
            this.requiredErrorText = 'The address is required';
            this.invalidErrorText = 'The address is not valid';
            this.appearance = Appearance.STANDARD;
            this.autoCompleteOptions = {};
            this.onChange = new core.EventEmitter();
            this.onAutocompleteSelected = new core.EventEmitter();
            this.onGermanAddressMapped = new core.EventEmitter();
            this.onLocationSelected = new core.EventEmitter();
            this.onNewPlaceResult = new core.EventEmitter();
            this.addressValidator = new MatValidateAddressDirective();
            this.addressSearchControl = new forms.FormControl({ value: null }, forms.Validators.compose([
                forms.Validators.required,
                this.addressValidator.validate()
            ]));
        }
        /**
         * @return {?}
         */
        MatGoogleMapsAutocompleteComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.addressValidator.subscribe(this.onNewPlaceResult);
            /** @type {?} */
            var options = {
                // types: ['address'],
                // componentRestrictions: {country: this.country},
                placeIdOnly: this.placeIdOnly,
                strictBounds: this.strictBounds,
                // types: this.types,
                type: this.type
            };
            // tslint:disable-next-line:no-unused-expression
            this.country ? options.componentRestrictions = { country: this.country } : null;
            // tslint:disable-next-line:no-unused-expression
            this.country ? options.types = this.types : null;
            this.autoCompleteOptions = Object.assign(this.autoCompleteOptions, options);
            this.initGoogleMapsAutocomplete();
        };
        /**
         * @return {?}
         */
        MatGoogleMapsAutocompleteComponent.prototype.initGoogleMapsAutocomplete = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.mapsAPILoader
                .load()
                .then((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var autocomplete = new google.maps.places.Autocomplete(_this.searchElementRef.nativeElement, _this.autoCompleteOptions);
                autocomplete.addListener('place_changed', (/**
                 * @return {?}
                 */
                function () {
                    _this.ngZone.run((/**
                     * @return {?}
                     */
                    function () {
                        // get the place result
                        /** @type {?} */
                        var place = autocomplete.getPlace();
                        /** @type {?} */
                        var germanAddress = {
                            gmID: place.id,
                            icon: place.icon,
                            url: place.url,
                            placeID: place.place_id,
                            displayAddress: place.formatted_address,
                            name: place.name,
                            vicinity: place.vicinity,
                            locality: {},
                            state: {},
                            country: {},
                            geoLocation: { latitude: -1, longitude: -1 },
                        };
                        if (place.geometry && place.geometry.location) {
                            germanAddress.geoLocation.latitude = place.geometry.location.lat();
                            germanAddress.geoLocation.longitude = place.geometry.location.lng();
                        }
                        place.address_components.forEach((/**
                         * @param {?} value
                         * @return {?}
                         */
                        function (value) {
                            if (value.types.indexOf('street_number') > -1) {
                                germanAddress.streetNumber = Number(value.short_name);
                            }
                            if (value.types.indexOf('route') > -1) {
                                germanAddress.streetName = value.long_name;
                            }
                            if (value.types.indexOf('postal_code') > -1) {
                                germanAddress.postalCode = Number(value.short_name);
                            }
                            if (value.types.indexOf('sublocality') > -1) {
                                germanAddress.sublocality = value.long_name;
                            }
                            if (value.types.indexOf('locality') > -1) {
                                germanAddress.locality.long = value.long_name;
                                germanAddress.locality.short = value.short_name;
                            }
                            if (value.types.indexOf('administrative_area_level_1') > -1) {
                                germanAddress.state.long = value.long_name;
                                germanAddress.state.short = value.short_name;
                            }
                            if (value.types.indexOf('country') > -1) {
                                germanAddress.country.long = value.long_name;
                                germanAddress.country.short = value.short_name;
                            }
                            if (value.types.indexOf('administrative_area_level_3') > -1) {
                                germanAddress.locality.short = value.short_name;
                            }
                        }));
                        _this.onGermanAddressMapped.emit(germanAddress);
                        if (!place.place_id || place.geometry === undefined || place.geometry === null) {
                            // place result is not valid
                            return;
                        }
                        else {
                            // show dialog to select a address from the input
                            // emit failed event
                        }
                        _this.address = place.formatted_address;
                        _this.onAutocompleteSelected.emit(place);
                        // console.log('onAutocompleteSelected -> ', place);
                        _this.onLocationSelected.emit({
                            latitude: place.geometry.location.lat(),
                            longitude: place.geometry.location.lng()
                        });
                    }));
                }));
            }))
                .catch((/**
             * @param {?} err
             * @return {?}
             */
            function (err) { return console.log(err); }));
        };
        /**
         * @param {?} event
         * @return {?}
         */
        MatGoogleMapsAutocompleteComponent.prototype.onQuery = /**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            // console.log('onChange()', event);
            this.onChange.emit(this.address);
        };
        /**
         * @private
         * @return {?}
         */
        MatGoogleMapsAutocompleteComponent.prototype.resetAddress = /**
         * @private
         * @return {?}
         */
        function () {
            this.address = null;
            this.addressSearchControl.updateValueAndValidity();
        };
        MatGoogleMapsAutocompleteComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'mat-google-maps-autocomplete',
                        exportAs: 'matGoogleMapsAutocomplete',
                        template: "<mat-form-field class=\"full-width\" [appearance]=\"appearance\">\n  <mat-label>{{addressLabelText}}</mat-label>\n  <input matInput\n         [(ngModel)]=\"address\"\n         (change)=\"onQuery($event)\"\n         placeholder=\"{{placeholderText}}\"\n         class=\"form-control\"\n         #search\n         MatValidateAddress\n         required>\n  <mat-error *ngIf=\"addressSearchControl.hasError('required')\">\n    {{requiredErrorText}}\n  </mat-error>\n  <mat-error *ngIf=\"addressSearchControl.hasError('validateAddress')\">\n    {{invalidErrorText}}\n  </mat-error>\n</mat-form-field>\n",
                        styles: [".full-width{width:100%}"]
                    }] }
        ];
        /** @nocollapse */
        MatGoogleMapsAutocompleteComponent.ctorParameters = function () { return [
            { type: core$1.MapsAPILoader },
            { type: core.NgZone }
        ]; };
        MatGoogleMapsAutocompleteComponent.propDecorators = {
            searchElementRef: [{ type: core.ViewChild, args: ['search', { static: false },] }],
            addressLabelText: [{ type: core.Input }],
            placeholderText: [{ type: core.Input }],
            requiredErrorText: [{ type: core.Input }],
            invalidErrorText: [{ type: core.Input }],
            appearance: [{ type: core.Input }],
            address: [{ type: core.Input }],
            country: [{ type: core.Input }],
            placeIdOnly: [{ type: core.Input }],
            strictBounds: [{ type: core.Input }],
            types: [{ type: core.Input }],
            type: [{ type: core.Input }],
            autoCompleteOptions: [{ type: core.Input }],
            onChange: [{ type: core.Output }],
            onAutocompleteSelected: [{ type: core.Output }],
            onGermanAddressMapped: [{ type: core.Output }],
            onLocationSelected: [{ type: core.Output }]
        };
        return MatGoogleMapsAutocompleteComponent;
    }());
    if (false) {
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.searchElementRef;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.addressLabelText;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.placeholderText;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.requiredErrorText;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.invalidErrorText;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.appearance;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.address;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.country;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.placeIdOnly;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.strictBounds;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.types;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.type;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.autoCompleteOptions;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.onChange;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.onAutocompleteSelected;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.onGermanAddressMapped;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.onLocationSelected;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteComponent.prototype.onNewPlaceResult;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteComponent.prototype.addressValidator;
        /** @type {?} */
        MatGoogleMapsAutocompleteComponent.prototype.addressSearchControl;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteComponent.prototype.mapsAPILoader;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteComponent.prototype.ngZone;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/helpers/parser.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @param {?} placeResult
     * @return {?}
     */
    function parseGermanAddress(placeResult) {
        /** @type {?} */
        var germanAddress = {
            gmID: placeResult.id,
            icon: placeResult.icon,
            url: placeResult.url,
            placeID: placeResult.place_id,
            displayAddress: placeResult.formatted_address,
            name: placeResult.name,
            vicinity: placeResult.vicinity,
            locality: {},
            state: {},
            country: {},
            geoLocation: { latitude: -1, longitude: -1 },
        };
        if (placeResult.geometry && placeResult.geometry.location) {
            germanAddress.geoLocation.latitude = placeResult.geometry.location.lat();
            germanAddress.geoLocation.longitude = placeResult.geometry.location.lng();
        }
        if (placeResult.address_components && placeResult.address_components.length > 0) {
            placeResult.address_components.forEach((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value.types.indexOf('street_number') > -1) {
                    germanAddress.streetNumber = Number(value.short_name);
                }
                if (value.types.indexOf('route') > -1) {
                    germanAddress.streetName = value.long_name;
                }
                if (value.types.indexOf('postal_code') > -1) {
                    germanAddress.postalCode = Number(value.short_name);
                }
                if (value.types.indexOf('sublocality') > -1) {
                    germanAddress.sublocality = value.long_name;
                }
                if (value.types.indexOf('locality') > -1) {
                    germanAddress.locality.long = value.long_name;
                    germanAddress.locality.short = value.short_name;
                }
                if (value.types.indexOf('administrative_area_level_1') > -1) {
                    germanAddress.state.long = value.long_name;
                    germanAddress.state.short = value.short_name;
                }
                if (value.types.indexOf('country') > -1) {
                    germanAddress.country.long = value.long_name;
                    germanAddress.country.short = value.short_name;
                }
                if (value.types.indexOf('administrative_area_level_3') > -1) {
                    germanAddress.locality.short = value.short_name;
                }
            }));
        }
        return germanAddress;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/animations/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var customAnimation = animations.animation([
        animations.style({
            opacity: '{{opacity}}',
            transform: 'scale({{scale}}) translate3d({{x}}, {{y}}, {{z}})'
        }),
        animations.animate('{{duration}} {{delay}} cubic-bezier(0.0, 0.0, 0.2, 1)', animations.style('*'))
    ], {
        params: {
            duration: '200ms',
            delay: '0ms',
            opacity: '0',
            scale: '1',
            x: '0',
            y: '0',
            z: '0'
        }
    });
    /** @type {?} */
    var InputAnimations = [
        animations.trigger('animate', [animations.transition('void => *', [animations.useAnimation(customAnimation)])]),
        animations.trigger('animateStagger', [
            animations.state('50', animations.style('*')),
            animations.state('100', animations.style('*')),
            animations.state('200', animations.style('*')),
            animations.transition('void => 50', animations.query('@*', [animations.stagger('50ms', [animations.animateChild()])], { optional: true })),
            animations.transition('void => 100', animations.query('@*', [animations.stagger('100ms', [animations.animateChild()])], { optional: true })),
            animations.transition('void => 200', animations.query('@*', [animations.stagger('200ms', [animations.animateChild()])], { optional: true }))
        ]),
    ];

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component/mat-search-google-maps-autocomplete/mat-search-google-maps-autocomplete.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatSearchGoogleMapsAutocompleteComponent = /** @class */ (function () {
        function MatSearchGoogleMapsAutocompleteComponent(formBuilder) {
            this.formBuilder = formBuilder;
            this.appearance = Appearance.STANDARD;
            this.searchAddressLabel = 'Search Address';
            this.streetNameLabel = 'Street';
            this.streetNumberLabel = 'Nr.';
            this.postalCodeLabel = 'PLZ';
            this.localityLabel = 'Locality';
            this.vicinityLabel = 'Vicinity';
            this.onGermanAddressMapped = new core.EventEmitter();
        }
        /**
         * @return {?}
         */
        MatSearchGoogleMapsAutocompleteComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.addressFormGroup = this.createAddressFormGroup();
        };
        /**
         * @return {?}
         */
        MatSearchGoogleMapsAutocompleteComponent.prototype.createAddressFormGroup = /**
         * @return {?}
         */
        function () {
            return this.formBuilder.group({
                streetName: [null, forms.Validators.required],
                streetNumber: [null, forms.Validators.required],
                postalCode: [null, forms.Validators.required],
                vicinity: [null],
                locality: this.formBuilder.group({
                    long: [null, forms.Validators.required],
                }),
            });
        };
        /**
         * @param {?} $event
         * @return {?}
         */
        MatSearchGoogleMapsAutocompleteComponent.prototype.syncAutoComplete = /**
         * @param {?} $event
         * @return {?}
         */
        function ($event) {
            if (this.germanAddress) {
                this.addressFormGroup.reset();
            }
            /** @type {?} */
            var germanAddress = parseGermanAddress($event);
            this.germanAddress = germanAddress;
            if (germanAddress.vicinity) {
                this.addressFormGroup.get('vicinity').patchValue(germanAddress.vicinity);
            }
            if (germanAddress.streetName) {
                this.addressFormGroup.get('streetName').patchValue(germanAddress.streetName);
            }
            if (germanAddress.streetNumber) {
                this.addressFormGroup.get('streetNumber').patchValue(germanAddress.streetNumber);
            }
            if (germanAddress.postalCode) {
                this.addressFormGroup.get('postalCode').patchValue(germanAddress.postalCode);
            }
            if (germanAddress.locality && germanAddress.locality.long) {
                this.addressFormGroup.get('locality.long').patchValue(germanAddress.locality.long);
            }
            this.onGermanAddressMapped.emit(germanAddress);
        };
        MatSearchGoogleMapsAutocompleteComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'mat-search-google-maps-autocomplete',
                        template: "<div fxLayout=\"column\">\n  <div fxFlex=\"100\">\n    <!--search address-->\n    <mat-form-field fxFlex=\"auto\" [appearance]=\"appearance\" [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n      <mat-label>{{searchAddressLabel}}</mat-label>\n      <input\n        (onAutocompleteSelected)=\"syncAutoComplete($event)\"\n        [country]=\"country\"\n        [placeIdOnly]=\"placeIdOnly\"\n        [strictBounds]=\"strictBounds\"\n        [types]=\"types\"\n        [type]=\"type\"\n        matGoogleMapsAutocomplete\n        matInput\n        required\n      />\n      <mat-icon color=\"primary\" matSuffix>search</mat-icon>\n      <!--    <mat-error>{{ 'msa.contactData.currentAddress.error' | translate }}</mat-error>-->\n    </mat-form-field>\n  </div>\n\n  <form [formGroup]=\"addressFormGroup\" fxFlex fxLayoutGap=\"10px\">\n    <div fxLayout=\"row\" fxLayoutGap=\"10px\">\n      <mat-form-field fxFlex=\"80\"\n                      [appearance]=\"appearance\"\n                      [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n        <mat-label>{{streetNameLabel}}</mat-label>\n        <input\n          [readonly]=\"readonly\"\n          formControlName=\"streetName\"\n          matInput\n          required\n        />\n        <!--        <mat-icon color=\"primary\" matSuffix>add_location</mat-icon>-->\n        <!--    <mat-error>{{ 'msa.contactData.currentAddress.error' | translate }}</mat-error>-->\n      </mat-form-field>\n      <mat-form-field fxFlex=\"20\" [appearance]=\"appearance\" [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n        <mat-label>{{streetNumberLabel}}</mat-label>\n        <input\n          [readonly]=\"readonly\"\n          formControlName=\"streetNumber\"\n          matInput\n          required\n        />\n        <!--        <mat-icon color=\"primary\" matSuffix>add_location</mat-icon>-->\n        <!--    <mat-error>{{ 'msa.contactData.currentAddress.error' | translate }}</mat-error>-->\n      </mat-form-field>\n    </div>\n    <div fxLayout=\"row\" fxLayoutGap=\"10px\">\n      <mat-form-field fxFlex=\"20\" [appearance]=\"appearance\" [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n        <mat-label>{{postalCodeLabel}}</mat-label>\n        <input\n          [readonly]=\"readonly\"\n          formControlName=\"postalCode\"\n          matInput\n          required\n        />\n        <!--        <mat-icon color=\"primary\" matSuffix>add_location</mat-icon>-->\n        <!--    <mat-error>{{ 'msa.contactData.currentAddress.error' | translate }}</mat-error>-->\n      </mat-form-field>\n      <mat-form-field *ngIf=\"showVicinity\" fxFlex=\"auto\"\n                      [appearance]=\"appearance\"\n                      [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n        <mat-label>{{vicinityLabel}}</mat-label>\n        <input\n          [readonly]=\"readonly\"\n          matInput\n          formControlName=\"vicinity\"\n        />\n      </mat-form-field>\n      <div formGroupName=\"locality\" fxFlex=\"auto\">\n        <mat-form-field fxFlex=\"auto\" [appearance]=\"appearance\" [@animate]=\"{ value: '*', params: { y: '100%' } }\">\n          <mat-label>{{localityLabel}}</mat-label>\n          <input\n            [readonly]=\"readonly\"\n            formControlName=\"long\"\n            matInput\n            required\n          />\n          <mat-icon color=\"primary\" matSuffix>add_location</mat-icon>\n          <!--    <mat-error>{{ 'msa.contactData.currentAddress.error' | translate }}</mat-error>-->\n        </mat-form-field>\n      </div>\n    </div>\n  </form>\n</div>\n",
                        animations: InputAnimations,
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        MatSearchGoogleMapsAutocompleteComponent.ctorParameters = function () { return [
            { type: forms.FormBuilder }
        ]; };
        MatSearchGoogleMapsAutocompleteComponent.propDecorators = {
            appearance: [{ type: core.Input }],
            searchAddressLabel: [{ type: core.Input }],
            streetNameLabel: [{ type: core.Input }],
            streetNumberLabel: [{ type: core.Input }],
            postalCodeLabel: [{ type: core.Input }],
            localityLabel: [{ type: core.Input }],
            vicinityLabel: [{ type: core.Input }],
            showVicinity: [{ type: core.Input }],
            country: [{ type: core.Input }],
            placeIdOnly: [{ type: core.Input }],
            strictBounds: [{ type: core.Input }],
            types: [{ type: core.Input }],
            type: [{ type: core.Input }],
            readonly: [{ type: core.Input }],
            disableSearch: [{ type: core.Input }],
            onGermanAddressMapped: [{ type: core.Output }]
        };
        return MatSearchGoogleMapsAutocompleteComponent;
    }());
    if (false) {
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.appearance;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.searchAddressLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.streetNameLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.streetNumberLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.postalCodeLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.localityLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.vicinityLabel;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.showVicinity;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.country;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.placeIdOnly;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.strictBounds;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.types;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.type;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.readonly;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.disableSearch;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.onGermanAddressMapped;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.germanAddress;
        /** @type {?} */
        MatSearchGoogleMapsAutocompleteComponent.prototype.addressFormGroup;
        /**
         * @type {?}
         * @private
         */
        MatSearchGoogleMapsAutocompleteComponent.prototype.formBuilder;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/mat-google-maps-autocomplete.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatGoogleMapsAutocompleteDirective = /** @class */ (function () {
        function MatGoogleMapsAutocompleteDirective(platformId, elemRef, mapsAPILoader, ngZone) {
            this.platformId = platformId;
            this.elemRef = elemRef;
            this.mapsAPILoader = mapsAPILoader;
            this.ngZone = ngZone;
            this.autoCompleteOptions = {};
            this.onChange = new core.EventEmitter();
            this.onAutocompleteSelected = new core.EventEmitter();
            this.onGermanAddressMapped = new core.EventEmitter();
            this.onLocationSelected = new core.EventEmitter();
            this.onNewPlaceResult = new core.EventEmitter();
            this.addressValidator = new MatValidateAddressDirective();
            this.addressSearchControl = new forms.FormControl({ value: null }, forms.Validators.compose([
                forms.Validators.required,
                this.addressValidator.validate()
            ]));
        }
        /**
         * @return {?}
         */
        MatGoogleMapsAutocompleteDirective.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            if (common.isPlatformBrowser(this.platformId)) {
                this.addressValidator.subscribe(this.onNewPlaceResult);
                /** @type {?} */
                var options = {
                    // types: ['address'],
                    // componentRestrictions: {country: this.country},
                    placeIdOnly: this.placeIdOnly,
                    strictBounds: this.strictBounds,
                    // types: this.types,
                    type: this.type
                };
                // tslint:disable-next-line:no-unused-expression
                this.country ? options.componentRestrictions = { country: this.country } : null;
                // tslint:disable-next-line:no-unused-expression
                this.country ? options.types = this.types : null;
                this.autoCompleteOptions = Object.assign(this.autoCompleteOptions, options);
                this.initGoogleMapsAutocomplete();
            }
        };
        /**
         * @return {?}
         */
        MatGoogleMapsAutocompleteDirective.prototype.initGoogleMapsAutocomplete = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.mapsAPILoader
                .load()
                .then((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var autocomplete = new google.maps.places.Autocomplete(_this.elemRef.nativeElement, _this.autoCompleteOptions);
                autocomplete.addListener('place_changed', (/**
                 * @return {?}
                 */
                function () {
                    _this.ngZone.run((/**
                     * @return {?}
                     */
                    function () {
                        // get the place result
                        /** @type {?} */
                        var place = autocomplete.getPlace();
                        /** @type {?} */
                        var germanAddress = {
                            gmID: place.id,
                            icon: place.icon,
                            url: place.url,
                            placeID: place.place_id,
                            displayAddress: place.formatted_address,
                            name: place.name,
                            vicinity: place.vicinity,
                            locality: {},
                            state: {},
                            country: {},
                            geoLocation: { latitude: -1, longitude: -1 },
                        };
                        if (place.geometry && place.geometry.location) {
                            germanAddress.geoLocation.latitude = place.geometry.location.lat();
                            germanAddress.geoLocation.longitude = place.geometry.location.lng();
                        }
                        place.address_components.forEach((/**
                         * @param {?} value
                         * @return {?}
                         */
                        function (value) {
                            if (value.types.indexOf('street_number') > -1) {
                                germanAddress.streetNumber = Number(value.short_name);
                            }
                            if (value.types.indexOf('route') > -1) {
                                germanAddress.streetName = value.long_name;
                            }
                            if (value.types.indexOf('postal_code') > -1) {
                                germanAddress.postalCode = Number(value.short_name);
                            }
                            if (value.types.indexOf('sublocality') > -1) {
                                germanAddress.sublocality = value.long_name;
                            }
                            if (value.types.indexOf('locality') > -1) {
                                germanAddress.locality.long = value.long_name;
                                germanAddress.locality.short = value.short_name;
                            }
                            if (value.types.indexOf('administrative_area_level_1') > -1) {
                                germanAddress.state.long = value.long_name;
                                germanAddress.state.short = value.short_name;
                            }
                            if (value.types.indexOf('country') > -1) {
                                germanAddress.country.long = value.long_name;
                                germanAddress.country.short = value.short_name;
                            }
                            if (value.types.indexOf('administrative_area_level_3') > -1) {
                                germanAddress.locality.short = value.short_name;
                            }
                        }));
                        _this.onGermanAddressMapped.emit(germanAddress);
                        if (!place.place_id || place.geometry === undefined || place.geometry === null) {
                            // place result is not valid
                            return;
                        }
                        else {
                            // show dialog to select a address from the input
                            // emit failed event
                        }
                        _this.address = place.formatted_address;
                        _this.onAutocompleteSelected.emit(place);
                        _this.onLocationSelected.emit({
                            latitude: place.geometry.location.lat(),
                            longitude: place.geometry.location.lng()
                        });
                    }));
                }));
            }))
                .catch((/**
             * @param {?} err
             * @return {?}
             */
            function (err) { return console.log(err); }));
        };
        MatGoogleMapsAutocompleteDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[matGoogleMapsAutocomplete]',
                        exportAs: 'matGoogleMapsAutocomplete',
                    },] }
        ];
        /** @nocollapse */
        MatGoogleMapsAutocompleteDirective.ctorParameters = function () { return [
            { type: String, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
            { type: core.ElementRef },
            { type: core$1.MapsAPILoader },
            { type: core.NgZone }
        ]; };
        MatGoogleMapsAutocompleteDirective.propDecorators = {
            address: [{ type: core.Input }],
            country: [{ type: core.Input }],
            placeIdOnly: [{ type: core.Input }],
            strictBounds: [{ type: core.Input }],
            types: [{ type: core.Input }],
            type: [{ type: core.Input }],
            autoCompleteOptions: [{ type: core.Input }],
            onChange: [{ type: core.Output }],
            onAutocompleteSelected: [{ type: core.Output }],
            onGermanAddressMapped: [{ type: core.Output }],
            onLocationSelected: [{ type: core.Output }]
        };
        return MatGoogleMapsAutocompleteDirective;
    }());
    if (false) {
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.address;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.country;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.placeIdOnly;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.strictBounds;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.types;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.type;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.autoCompleteOptions;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.onChange;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.onAutocompleteSelected;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.onGermanAddressMapped;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.onLocationSelected;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteDirective.prototype.onNewPlaceResult;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteDirective.prototype.addressValidator;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.addressSearchControl;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.platformId;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.elemRef;
        /** @type {?} */
        MatGoogleMapsAutocompleteDirective.prototype.mapsAPILoader;
        /**
         * @type {?}
         * @private
         */
        MatGoogleMapsAutocompleteDirective.prototype.ngZone;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/mat-google-maps-autocomplete.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatGoogleMapsAutocompleteModule = /** @class */ (function () {
        function MatGoogleMapsAutocompleteModule() {
        }
        MatGoogleMapsAutocompleteModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            forms.ReactiveFormsModule,
                            flexLayout.FlexLayoutModule,
                            material.MatInputModule,
                            material.MatIconModule
                        ],
                        exports: [
                            MatGoogleMapsAutocompleteComponent,
                            MatGoogleMapsAutocompleteDirective,
                            MatValidateAddressDirective,
                            MatSearchGoogleMapsAutocompleteComponent
                        ],
                        declarations: [
                            MatGoogleMapsAutocompleteComponent,
                            MatGoogleMapsAutocompleteDirective,
                            MatValidateAddressDirective,
                            MatSearchGoogleMapsAutocompleteComponent
                        ]
                    },] }
        ];
        return MatGoogleMapsAutocompleteModule;
    }());

    exports.Appearance = Appearance;
    exports.MatGoogleMapsAutocompleteComponent = MatGoogleMapsAutocompleteComponent;
    exports.MatGoogleMapsAutocompleteDirective = MatGoogleMapsAutocompleteDirective;
    exports.MatGoogleMapsAutocompleteModule = MatGoogleMapsAutocompleteModule;
    exports.MatSearchGoogleMapsAutocompleteComponent = MatSearchGoogleMapsAutocompleteComponent;
    exports.MatValidateAddressDirective = MatValidateAddressDirective;
    exports.ɵa = MatGoogleMapsAutocompleteComponent;
    exports.ɵb = MatSearchGoogleMapsAutocompleteComponent;
    exports.ɵc = InputAnimations;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
